pyprecag/processing.py

```diff
@@ -2537,3 +2401,222 @@ def ttest_analysis(points_geodataframe, points_crs, values_raster, out_folder,
+                   a integer between -50 and 50
+
+    Returns:
+        str: The output tif name
+    """
+
+    if not isinstance(target_percentage, int):
+        raise TypeError('target_percentage must an integer between -50 to 50.')
+
+    if target_percentage < -50 or target_percentage > 50:
+        raise ValueError('target_percentage must an integer between -50 to 50.')
+
+    if not isinstance(greater_than, (int, bool)):
+        raise TypeError('greater_than must be boolean.')
+
+    if output_tif is not None and not os.path.isabs(output_tif):
+        output_tif = os.path.join(TEMPDIR, output_tif)
+
+    if not isinstance(raster_files, list):
+        raise TypeError('Invalid Type: raster_files should be a list')
+
+    if len(raster_files) == 0:
+        raise TypeError('Invalid Type: Empty list of raster files')
+
+    start_time = time.time()
+
+    # combine and align grid and slip to common area
+    image_combined, _ = stack_and_clip_rasters(raster_files, use_common=True,
+                                               output_tif='allyrs_combined.tif')
+
+    with rasterio.open(image_combined, 'r') as src:
+        kwargs = src.meta.copy()
+
+        for i, band in enumerate(src.read(masked=True), start=1):
+            cutoff = np.nanmean(band) + (np.nanmean(band) * (target_percentage / 100.00))
+            band_new = band.copy()
+
+            # classifiy bands
+            band_new[np.where(band >= cutoff)] = (greater_than == True)
+            band_new[np.where(band < cutoff)] = (greater_than != True)
+
+            # reapply nodata values
+            band_new[np.where(band.mask)] = -9999
+            band_new = band_new.astype(np.int16)
+
+            # cumulative sum of bands
+            if i == 1:
+                sum_of_rasters = band_new
+            else:
+                sum_of_rasters = sum_of_rasters + band_new
+
+            sum_of_rasters[band.mask] = -9999
+
+            del band, band_new
+
+    kwargs.pop('count')
+    kwargs.update({'nodata': -9999, 'dtype': np.int16})
+    with rasterio.open(output_tif, 'w', count=1, **kwargs) as dst_allyrs:
+        dst_allyrs.write(sum_of_rasters.astype(np.int16), 1)
+
+    arg_str = '{} {}%'.format('>' if greater_than else '<', target_percentage)
+    LOGGER.info('{:<30} {:>15} {dur}'.format('Persistor All Years Completed', arg_str,
+                                             dur=timedelta(seconds=time.time() - start_time)))
+
+    # cleanup Temp files
+    if not config.get_debug_mode() and TEMPDIR in image_combined:
+        os.remove(image_combined)
+
+    return output_tif
+
+
+def persistor_target_probability(upper_raster_files, upper_percentage, upper_probability,
+                                 lower_raster_files, lower_percentage, lower_probability,
+                                 output_tif):
+    """Determine the probability of a performance being exceeded or not being met as described in
```

Comment:
As we both realise, this is difficult to summarise for an unfamiliar user. I suggest rewording as follows ... 

"
The "Target probability" method analysis and combines two categories (UPPER and LOWER) of rasters and rules to 
determine which cells meet a higher performance rule (Upper category), or a lower performance rule (Lower 
category) as described in Bramley and Hamilton (2005). This is typically applied to data such as yield maps
with one yield raster per year. 

Each pixel of the output raster is assigned a value as follows:
    1 if the upper category rasters meet the upper category rule, 
    -1 if the lower category rasters meet the lower category rule
    zero otherwise

The upper and lower category rules are defined as follows: 
    Upper: 
        a threshold for each raster calculated from the mean of each raster +/- a specified percentage  
        a proportion of the upper category rasters which must be ABOVE the threshold 
    Lower: 
        a threshold for each raster calculated from the mean of each raster +/- a specified percentage  
        a proportion of the lower category rasters which must be BELOW the threshold

Reply:
Updated.